assets/js/blocks/product-query/variations/product-query.tsx

```diff
@@ -19,33 +20,56 @@ import {
 
 const VARIATION_NAME = 'woocommerce/product-query';
 
+const ARCHIVE_PRODUCT_TEMPLATES = [
+	'woocommerce/woocommerce//archive-product',
+	'woocommerce/woocommerce//taxonomy-product_cat',
+	'woocommerce/woocommerce//taxonomy-product_tag',
+	'woocommerce/woocommerce//taxonomy-product_attribute',
+	'woocommerce/woocommerce//product-search-results',
+];
+
 if ( isWpVersion( '6.1', '>=' ) ) {
-	registerBlockVariation( QUERY_LOOP_ID, {
-		description: __(
-			'A block that displays a selection of products in your store.',
-			'woo-gutenberg-products-block'
-		),
-		name: VARIATION_NAME,
-		/* translators: “Products“ is the name of the block. */
-		title: __( 'Products (Beta)', 'woo-gutenberg-products-block' ),
-		isActive: ( blockAttributes ) =>
-			blockAttributes.namespace === VARIATION_NAME,
-		icon: (
-			<Icon
-				icon={ stacks }
-				className="wc-block-editor-components-block-icon wc-block-editor-components-block-icon--stacks"
-			/>
-		),
-		attributes: {
-			...QUERY_DEFAULT_ATTRIBUTES,
-			namespace: VARIATION_NAME,
-		},
-		// Gutenberg doesn't support this type yet, discussion here:
-		// https://github.com/WordPress/gutenberg/pull/43632
-		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-		// @ts-ignore
-		allowedControls: DEFAULT_ALLOWED_CONTROLS,
-		innerBlocks: INNER_BLOCKS_TEMPLATE,
-		scope: [ 'inserter' ],
+	let currentTemplateId: string | undefined;
+
+	subscribe( () => {
+		const previousTemplateId = currentTemplateId;
+		const store = select( 'core/edit-site' );
+		currentTemplateId = store?.getEditedPostId();
+
+		if ( previousTemplateId === currentTemplateId ) {
+			return;
+		}
+
+		QUERY_DEFAULT_ATTRIBUTES.query.inherit =
+			ARCHIVE_PRODUCT_TEMPLATES.includes( currentTemplateId );
+
+		registerBlockVariation( QUERY_LOOP_ID, {
+			description: __(
+				'A block that displays a selection of products in your store.',
+				'woo-gutenberg-products-block'
+			),
+			name: VARIATION_NAME,
+			/* translators: “Products“ is the name of the block. */
+			title: __( 'Products (Beta)', 'woo-gutenberg-products-block' ),
+			isActive: ( blockAttributes ) =>
+				blockAttributes.namespace === VARIATION_NAME,
+			icon: (
+				<Icon
+					icon={ stacks }
+					className="wc-block-editor-components-block-icon wc-block-editor-components-block-icon--stacks"
+				/>
+			),
+			attributes: {
+				...QUERY_DEFAULT_ATTRIBUTES,
+				namespace: VARIATION_NAME,
+			},
+			// Gutenberg doesn't support this type yet, discussion here:
+			// https://github.com/WordPress/gutenberg/pull/43632
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore
+			allowedControls: DEFAULT_ALLOWED_CONTROLS,
+			innerBlocks: INNER_BLOCKS_TEMPLATE,
+			scope: [ 'inserter' ],
+		} );
```

Comment:
Minor: please consider extracting `registerBLockVariation` to a separate function like `const registerProducts = (attributes) => { ... }`. I suggest this to improve readability of `subscibe`'s listener. At first sight it looks more scary than it really is 😅

Reply:
Done! 👍